daisy/workflow/workflow.go

```diff
@@ -315,26 +308,27 @@ func (w *Workflow) FromFile(file string) error {
 // Run runs a workflow.
 func (w *Workflow) Run() error {
 	if err := w.populate(); err != nil {
-		w.Cancel()
+		close(w.Cancel)
 		return err
 	}
 
 	w.logger.Print("Validating workflow")
 	if err := w.validate(); err != nil {
 		w.logger.Printf("Error validating workflow: %v", err)
-		w.Cancel()
+		close(w.Cancel)
 		return err
 	}
 	w.logger.Print("Uploading sources")
 	if err := w.uploadSources(); err != nil {
 		w.logger.Printf("Error uploading sources: %v", err)
-		w.Cancel()
+		close(w.Cancel)
 		return err
 	}
 	w.logger.Print("Running workflow")
+	defer w.cleanup()
```

Comment:
We could maybe even put this at the top of Run(). Say, for example, we want to clean up artifacts from uploadSources().